sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashAggregateExec.scala

```diff
@@ -244,6 +246,92 @@ case class HashAggregateExec(
 
   protected override val shouldStopRequired = false
 
+  // We assume a prefix has lower cases and a name has camel cases
+  private val variableName = "^[a-z]+_[a-zA-Z]+[0-9]*".r
+
+  // Returns true if a given name id belongs to this `CodegenContext`
+  private def isVariable(nameId: String): Boolean = nameId match {
+    case variableName() => true
+    case _ => false
+  }
+
+  // Extracts all the outer references for a given `aggExpr`. This result will be used to split
+  // aggregation into small functions.
+  private def getOuterReferences(
+      ctx: CodegenContext,
+      aggExpr: Expression,
+      subExprs: Map[Expression, SubExprEliminationState]): Set[(String, String)] = {
+    val stack = mutable.Stack[Expression](aggExpr)
+    val argSet = mutable.Set[(String, String)]()
+    val addIfNotLiteral = (value: String, tpe: String) => {
```

Comment:
Hmm. Just a cosmetic style comment: I would have declared `addIfNotLiteral ` with a `def` instead of making it a `scala.Function2[String, String, Unit]`.

BTW, can we add a comment to `val argSet` for what those two fields of the `Tuple2[String, String]` means? And then also make this `addIfNotLiteral` function take the arguments in the same order as the tuple.